File: apps/ts/packages/api/src/services/market/market-data-service.ts
/**
 * Market Data Service
 * Provides data access for market.db
 * Used by Python API clients (trading25-bt)
 */
import * as fs from 'node:fs';
import { DrizzleMarketDataReader } from '@trading25/shared/db';
import { getMarketDbPath } from '@trading25/shared/utils/dataset-paths';
import { logger } from '@trading25/shared/utils/logger';
import type {
  MarketDateRangeQuery,
  MarketOHLCRecord,
  MarketOHLCVRecord,
  MarketStockData,
  MarketStocksQuery,
  StockInfo,
} from '../../schemas/market-data';

/**
 * Format date to YYYY-MM-DD string
 */
function formatDate(date: Date): string {
  return date.toISOString().split('T')[0] ?? '';
}

/**
 * Parse date string to Date object with validation
 */
function parseDate(dateStr: string | undefined): Date | undefined {
  if (!dateStr) return undefined;
  const date = new Date(dateStr);
  if (Number.isNaN(date.getTime())) {
    throw new Error(`Invalid date string: ${dateStr}`);
  }
  return date;
}

/**
 * Build and validate date range from query parameters
 */
function buildDateRange(query: {
  start_date?: string;
  end_date?: string;
}): { from: Date; to: Date } | undefined {
  if (!query.start_date && !query.end_date) {
    return undefined;
  }

  const from = parseDate(query.start_date) ?? new Date('1900-01-01');
  const to = parseDate(query.end_date) ?? new Date();

  if (from > to) {
    logger.warn('Invalid date range: start_date is after end_date', {
      start_date: query.start_date,
      end_date: query.end_date,
    });
  }

  return { from, to };
}

/**
 * Get market database reader (lazy singleton)
 */
let marketDbReaderInstance: DrizzleMarketDataReader | null = null;

function getReader(): DrizzleMarketDataReader | null {
  if (marketDbReaderInstance !== null) {
    return marketDbReaderInstance;
  }

  const dbPath = getMarketDbPath();
  if (!fs.existsSync(dbPath)) {
    logger.error('Market database not found', { dbPath });
    return null;
  }

  try {
    marketDbReaderInstance = new DrizzleMarketDataReader(dbPath);
    logger.info('Market database reader initialized', { dbPath });
    return marketDbReaderInstance;
  } catch (error) {
    logger.error('Failed to initialize market database reader', {
      dbPath,
      error: error instanceof Error ? error.message : error,
    });
    return null;
  }
}

/**
 * Get market code list from query parameter
 */
function getMarketCodes(market: 'prime' | 'standard'): string[] {
  return [market];
}

/**
 * Market Data Service
 */
export const marketDataService = {
  /**
   * Normalize stock code to 5-digit format for market.db lookup
   * market.db uses 5-digit codes with trailing "0" (e.g., "72030" for Toyota "7203")
   */
  normalizeStockCode(stockCode: string): string {
    return stockCode.length === 4 ? `${stockCode}0` : stockCode;
  },

  /**
   * Get single stock info by code
   */
  getStockInfo(stockCode: string): StockInfo | null {
    const reader = getReader();
    if (!reader) return null;

    try {
      const dbSymbol = this.normalizeStockCode(stockCode);
      const stockInfo = reader.getStockByCode(dbSymbol);
      if (!stockInfo) return null;

      return {
        code: stockInfo.code,
        companyName: stockInfo.companyName,
        companyNameEnglish: stockInfo.companyNameEnglish,
        marketCode: stockInfo.marketCode,
        marketName: stockInfo.marketName,
        sector17Code: stockInfo.sector17Code,
        sector17Name: stockInfo.sector17Name,
        sector33Code: stockInfo.sector33Code,
        sector33Name: stockInfo.sector33Name,
        scaleCategory: stockInfo.scaleCategory,
        listedDate: formatDate(stockInfo.listedDate),
      };
    } catch (error) {
      logger.error('Failed to get stock info from market.db', { stockCode, error });
      return null;
    }
  },

  /**
   * Get stock OHLCV data
   */
  getStockOHLCV(stockCode: string, query: MarketDateRangeQuery): MarketOHLCVRecord[] | null {
    const reader = getReader();
    if (!reader) return null;

    try {
      const dbSymbol = this.normalizeStockCode(stockCode);
      const dateRange = buildDateRange(query);

      const data = reader.getStockData(dbSymbol, dateRange);

      return data.map((d) => ({
        date: formatDate(d.date),
        open: d.open,
        high: d.high,
        low: d.low,
        close: d.close,
        volume: d.volume,
      }));
    } catch (error) {
      logger.error('Failed to get stock OHLCV from market.db', { stockCode, error });
      return null;
    }
  },

  /**
   * Get all stocks data for screening
   * Returns stock list with OHLCV history for specified period
   */
  getAllStocks(query: MarketStocksQuery): MarketStockData[] | null {
    const reader = getReader();
    if (!reader) return null;

    try {
      const marketCodes = getMarketCodes(query.market);
      const stocks = reader.getStockList(marketCodes);

      // Calculate date range based on history_days
      const latestDate = reader.getLatestTradingDate();
      if (!latestDate) {
        logger.warn('No trading data found in market.db');
        return [];
      }

      const startDate = reader.getTradingDateBefore(latestDate, query.history_days);
      const dateRange = startDate
        ? { from: startDate, to: latestDate }
        : { from: new Date('1900-01-01'), to: latestDate };

      const result: MarketStockData[] = [];

      for (const stock of stocks) {
        const stockData = reader.getStockData(stock.code, dateRange);
        if (stockData.length > 0) {
          result.push({
            code: stock.code,
            company_name: stock.companyName,
            data: stockData.map((d) => ({
              date: formatDate(d.date),
              open: d.open,
              high: d.high,
              low: d.low,
              close: d.close,
              volume: d.volume,
            })),
          });
        }
      }

      logger.debug('Retrieved all stocks data from market.db', {
        market: query.market,
        stockCount: result.length,
        historyDays: query.history_days,
      });

      return result;
    } catch (error) {
      logger.error('Failed to get all stocks from market.db', { error });
      return null;
    }
  },

  /**
   * Get TOPIX data
   */
  getTopix(query: MarketDateRangeQuery): MarketOHLCRecord[] | null {
    const reader = getReader();
    if (!reader) return null;

    try {
      const dateRange = buildDateRange(query);

      const data = reader.getTopixData(dateRange);

      return data.map((d) => ({
        date: formatDate(d.date),
        open: d.open,
        high: d.high,
        low: d.low,
        close: d.close,
      }));
    } catch (error) {
      logger.error('Failed to get TOPIX data from market.db', { error });
      return null;
    }
  },
};

File: apps/bt/src/server/services/market_data_service.py
"""
Market Data Service

market.db から株式・TOPIX データを読み取るサービス。
Hono market-data-service.ts と同等のロジック。
"""

from __future__ import annotations

from src.lib.market_db.market_reader import MarketDbReader
from src.lib.market_db.query_helpers import stock_code_candidates
from src.server.services.market_code_alias import resolve_market_codes
from src.server.schemas.market_data import (
    MarketOHLCRecord,
    MarketOHLCVRecord,
    MarketStockData,
    StockInfo,
)


def _stock_code_candidates(code: str) -> tuple[str, ...]:
    """DB検索用の銘柄コード候補（4桁/5桁両対応）"""
    return stock_code_candidates(code)


class MarketDataService:
    """market.db 読み取りサービス"""

    def __init__(self, reader: MarketDbReader) -> None:
        self._reader = reader

    @staticmethod
    def _coerce_volume(value: object) -> int:
        """Convert DB volume value to int with safe fallback for null/invalid values."""
        if value is None:
            return 0
        try:
            return int(value)
        except (TypeError, ValueError):
            try:
                return int(float(str(value)))
            except (TypeError, ValueError):
                return 0

    def get_stock_info(self, code: str) -> StockInfo | None:
        """単一銘柄の情報を取得"""
        codes = _stock_code_candidates(code)
        placeholders = ",".join("?" for _ in codes)
        row = self._reader.query_one(
            "SELECT code, company_name, company_name_english, market_code, market_name, "
            "sector_17_code, sector_17_name, sector_33_code, sector_33_name, "
            f"scale_category, listed_date FROM stocks WHERE code IN ({placeholders}) "
            "ORDER BY CASE WHEN length(code) = 4 THEN 0 ELSE 1 END LIMIT 1",
            tuple(codes),
        )
        if row is None:
            return None

        return StockInfo(
            code=row["code"],
            companyName=row["company_name"],
            companyNameEnglish=row["company_name_english"] or "",
            marketCode=row["market_code"],
            marketName=row["market_name"],
            sector17Code=row["sector_17_code"],
            sector17Name=row["sector_17_name"],
            sector33Code=row["sector_33_code"],
            sector33Name=row["sector_33_name"],
            scaleCategory=row["scale_category"] or "",
            listedDate=row["listed_date"],
        )

    def get_stock_ohlcv(
        self,
        code: str,
        start_date: str | None = None,
        end_date: str | None = None,
    ) -> list[MarketOHLCVRecord] | None:
        """銘柄の OHLCV データを取得"""
        codes = _stock_code_candidates(code)
        placeholders = ",".join("?" for _ in codes)

        # 銘柄存在確認
        row = self._reader.query_one(
            f"SELECT code FROM stocks WHERE code IN ({placeholders}) "
            "ORDER BY CASE WHEN length(code) = 4 THEN 0 ELSE 1 END LIMIT 1",
            tuple(codes),
        )
        if row is None:
            return None
        db_code = row["code"]

        sql = "SELECT date, open, high, low, close, volume FROM stock_data WHERE code = ?"
        params: list[str] = [db_code]

        if start_date:
            sql += " AND date >= ?"
            params.append(start_date)
        if end_date:
            sql += " AND date <= ?"
            params.append(end_date)

        sql += " ORDER BY date"

        rows = self._reader.query(sql, tuple(params))
        return [
            MarketOHLCVRecord(
                date=row["date"],
                open=row["open"],
                high=row["high"],
                low=row["low"],
                close=row["close"],
                volume=self._coerce_volume(row["volume"]),
            )
            for row in rows
        ]

    def get_all_stocks(
        self,
        market: str = "prime",
        history_days: int = 300,
    ) -> list[MarketStockData] | None:
        """市場別全銘柄データを取得（スクリーニング用）"""
        _, query_market_codes = resolve_market_codes(market, fallback=["prime"])

        # 最新取引日を取得
        latest_row = self._reader.query_one("SELECT MAX(date) as max_date FROM stock_data")
        if latest_row is None or latest_row["max_date"] is None:
            return []

        latest_date = latest_row["max_date"]

        # history_days 日前の取引日を取得
        start_row = self._reader.query_one(
            "SELECT DISTINCT date FROM stock_data WHERE date < ? ORDER BY date DESC LIMIT 1 OFFSET ?",
            (latest_date, history_days - 1),
        )
        start_date = start_row["date"] if start_row else "1900-01-01"

        # 銘柄一覧
        placeholders = ",".join("?" for _ in query_market_codes)
        stock_rows = self._reader.query(
            f"SELECT code, company_name FROM stocks WHERE market_code IN ({placeholders}) ORDER BY code",
            tuple(query_market_codes),
        )
        if not stock_rows:
            return None

        result: list[MarketStockData] = []
        for stock in stock_rows:
            data_rows = self._reader.query(
                "SELECT date, open, high, low, close, volume FROM stock_data "
                "WHERE code = ? AND date >= ? AND date <= ? ORDER BY date",
                (stock["code"], start_date, latest_date),
            )
            if data_rows:
                result.append(
                    MarketStockData(
                        code=stock["code"],
                        company_name=stock["company_name"],
                        data=[
                            MarketOHLCVRecord(
                                date=r["date"],
                                open=r["open"],
                                high=r["high"],
                                low=r["low"],
                                close=r["close"],
                                volume=self._coerce_volume(r["volume"]),
                            )
                            for r in data_rows
                        ],
                    )
                )

        return result

    def get_topix(
        self,
        start_date: str | None = None,
        end_date: str | None = None,
    ) -> list[MarketOHLCRecord] | None:
        """TOPIX データを取得"""
        sql = "SELECT date, open, high, low, close FROM topix_data"
        params: list[str] = []
        conditions: list[str] = []

        if start_date:
            conditions.append("date >= ?")
            params.append(start_date)
        if end_date:
            conditions.append("date <= ?")
            params.append(end_date)

        if conditions:
            sql += " WHERE " + " AND ".join(conditions)
        sql += " ORDER BY date"

        rows = self._reader.query(sql, tuple(params))
        if not rows:
            return None

        return [
            MarketOHLCRecord(
                date=row["date"],
                open=row["open"],
                high=row["high"],
                low=row["low"],
                close=row["close"],
            )
            for row in rows
        ]
